### PR TITLE
React Add participants: Checkbox cell for boolean headers

### LINKED ISSUE #157: Add participants: Checkbox cell for boolean headers (solved, affected)
Task points: 3

### PR DESCRIPTION
**Pull request by frewmack**
Closes #157

### CODE REVIEW

File: react/src/pages/utils/upload/DataEntryTable.tsx [hunk 1]
```diff
@@ -239,14 +249,51 @@ export default function DataEntryTable(props: DataEntryTableProps) {
...
+                value={toOption(props.row[props.col.field])}
+                options={props.getOptions(props.rowIndex, props.col)}
+                onEdit={props.onEdit}
+                aria-label={`enter ${props.col.title} row ${props.rowIndex}`}
+                disabled={props.disabled}
+            />
+        );
+    }
+
+    return cell;
```
Comment (kevinlul) [inline react/src/pages/utils/upload/DataEntryTable.tsx:286]: I think you can just return out of each branch

File: react/src/pages/utils/upload/DataEntryTable.tsx [hunk 2]
```diff
@@ -239,14 +249,51 @@ export default function DataEntryTable(props: DataEntryTableProps) {
...
+    getOptions: (rowIndex: number, col: DataEntryHeader) => Option[];
+    onEdit: (newValue: string | boolean) => void;
+    disabled?: boolean;
+}) {
+    let cell;
+
+    if (booleanColumns.includes(props.col.field)) {
+        cell = (
+            <CheckboxCell
+                value={props.row[props.col.field] as boolean}
```
Comment (kevinlul) [inline react/src/pages/utils/upload/DataEntryTable.tsx:269]:
```suggestion
                value={!!props.row[props.col.field]}
```

File: react/src/pages/utils/upload/DataEntryTable.tsx [hunk 3]
```diff
@@ -313,6 +360,25 @@ function DataEntryCell(
...
+    onEdit: (newValue: boolean) => void;
+    disabled?: boolean;
+}) {
+    function handleChange(event: React.ChangeEvent<HTMLInputElement>) {
+        props.onEdit(event.target.checked);
+    }
+
+    return (
+        <TableCell padding="checkbox">
+            <Checkbox checked={!!props.value} onChange={handleChange} />
```
Comment (kevinlul) [inline react/src/pages/utils/upload/DataEntryTable.tsx:377]:
```suggestion
            <Checkbox checked={props.value} onChange={event => props.onEdit(event.target.checked)} />
```
This should shorten this component by a lot. `props.disabled` isn't used at all though.

**Review by kevinlul — CHANGES_REQUESTED** [review]

File: react/src/pages/utils/upload/DataEntryTable.tsx [hunk 4]
```diff
@@ -1,6 +1,8 @@
 import React, { ReactNode, useEffect, useState } from "react";
 import {
     Box,
+    Button,
```
Comment (kevinlul) [inline react/src/pages/utils/upload/DataEntryTable.tsx:4]: This doesn't seem to be used anywhere. If it's needed for later then it's fine.

**Review by kevinlul — APPROVED** [review]
LGTM

**Review by annnag — COMMENTED** [review]
Looks good! One thing I noticed is that there isn't a state/icon to represent null, so I wonder if it is necessary to have one? Or having true and false is enough for the users?

**Comment by kevinlul** [comment]
There is an `indeterminate` state for mui checkboxes but I don't know how usable it is from a UX perspective.

**Comment by frewmack** [comment]
Right now the state just assumes that `null`/`undefined` is the same as `false`. The `indeterminate` state for checkboxes is usually used on nested checklist when only some of the 'child' checkboxes are checked, so I'm a little worried that users may find it confusing. Example: 
image

**Comment by helrick** [comment]
Would it be possible to convert the nullable boolean fields to dropdowns with a "Blank" option? In that case we could give them the three options that the database has and don't have to decide on a default value for these fields

**Comment by frewmack** [comment]
> Would it be possible to convert the nullable boolean fields to dropdowns with a "Blank" option? In that case we could give them the three options that the database has and don't have to decide on a default value for these fields

I can add this as an additional component so that we can compare in the next standup. I'll also make a copy of the checkbox component where `indeterminate` state is used to see how it looks.

**Review by annnag — APPROVED** [review]

**Pull request merged**